自動更新機能の実装

hidetoshi1230 opened the pull request:
WHAT
自動更新機能の実装

WHY
ユーザビリティ向上のため

hidetoshi1230 committed:
自動更新機能の実装１（実装）

定期的に非同期通信を行い、該当グループのメッセージを取得する
取得したメッセージでメッセージのHTMLを組み立てる
ビューに追加する

hidetoshi1230 commented:
テスト
https://gyazo.com/7c1e25acad8818550304fc63431f5182

hidollara commented on app/assets/javascripts/message.js:
```diff
@@ -43,4 +44,28 @@ $(function(){
+      .done(function(data) {
+        var id = $('.message').last().data('message-id');
+        data.forEach(function(message){
+          if (message.id > id ) {
```
全てのメッセージを引っ張ってきて一部のメッセージを使うのは、通信量がかさみます！

元から必要なメッセージだけを引っ張ってこれるようrails側で処理してしまうほうが良いです

hidetoshi1230 committed:
自動更新機能の実装２（修正）

全てのメッセージを引っ張ってきて一部のメッセージを使うのは、通信量がかさみます！
元から必要なメッセージだけを引っ張ってこれるようrails側で処理してしまうほうが良いです

hidetoshi1230 committed:
自動更新機能の実装３（修正）

コンソールログの削除

mkoma414 commented:
前回のコメントの内容がうまく伝わっていないかもしれないため
改めて修正していただきたい内容をお伝えいたします。

例えば今index画面で、idが100番までのメッセージが表示されているとします。

その際、最新のidが100であることをAjaxで送信して、
コントローラー側ではidが100より大きい（新しい）メッセージだけを
JSONとして返す、という動作を想定しています。

つまりデータのやり取りとしては新しいものだけが送信されれば良いのに
idが1番からのデータを全て送るとトラフィックが無駄になってしまう、
ということになります。

上記の動作をするようコードの修正をお願いします。

hidetoshi1230 committed:
自動更新機能の実装３（修正）

前回のコメントの内容がうまく伝わっていないかもしれないため
改めて修正していただきたい内容をお伝えいたします。

例えば今index画面で、idが100番までのメッセージが表示されているとします。

その際、最新のidが100であることをAjaxで送信して、
コントローラー側ではidが100より大きい（新しい）メッセージだけを
JSONとして返す、という動作を想定しています。

つまりデータのやり取りとしては新しいものだけが送信されれば良いのに
idが1番からのデータを全て送るとトラフィックが無駄になってしまう、
ということになります。

上記の動作をするようコードの修正をお願いします。

hidetoshi1230 commented:
下記のエラーがでてしまいます、ご確認のほど、よろしくお願いいたします。
wrong number of arguments (given 1, expected 0)

yamamotokeishiro commented on app/controllers/messages_controller.rb:
```diff
@@ -5,10 +5,13 @@ class MessagesController < ApplicationController
-      format.html
-      format.json
+    @messages.each do |message|
+      if message.id > params(:last_id)
```
この方法ですと、最新のメッセージが複数あった際に一つ一つjs側に返しています。
最新のメッセージを配列の形で得しjsに返しましょう！

ヒントはwhere文を使用して取得するタイミングで条件にあった値だけ取得します！

hidetoshi1230 committed:
自動更新機能の実装４

この方法ですと、最新のメッセージが複数あった際に一つ一つjs側に返しています。
最新のメッセージを配列の形で得しjsに返しましょう！

ヒントはwhere文を使用して取得するタイミングで条件にあった値だけ取得します！

smtrdev commented on app/views/messages/index.json.jbuilder:
```diff
@@ -0,0 +1,7 @@
+json.date     message.created_at.strftime("%Y-%m-%d %H:%M:%S")
+json.body     message.body
+json.image    message.image
+json.id       message.id
```
できれば2~6行目はインデントを下げましょう

smtrdev reviewed (commented):
インデント直して頂ければ、LGTMです！

hidetoshi1230 committed:
自動更新機能の実装５（修正）

できれば2~6行目はインデントを下げましょう
